File: src/utils/api_requests.py
##############################################
# Title: Modular API Request Script
# Author: Christopher Romanillos
# Description: modular api_request script
# Date: 11/23/24
# Version: 1.0
##############################################
import requests
import logging

def fetch_api_data(url, timeout):
    """Send a GET request to the API and return the data."""
    try:
        response = requests.get(url, timeout=timeout)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.Timeout:
        logging.error(f"Request timed out after {timeout} seconds.")
        raise
    except requests.exceptions.ConnectionError:
        logging.error("A connection error occurred.")
        raise
    except requests.exceptions.HTTPError as http_err:
        logging.error(f"HTTP error occurred: {http_err}")
        raise
    except requests.exceptions.RequestException as err:
        logging.error(f"An unexpected error occurred: {err}")
        raise

File: src/utils/api_requests.ts
/**
 * Title: Modular API Request Script
 * Author: Christopher Romanillos
 * Description: modular .ts api_request script
 * Date: 12/18/24
 * Version: 1.0
 */

import axios, { AxiosError } from 'axios';

export async function fetchApiData(url: string, timeout: number): Promise<any> {
    try {
        const response = await axios.get(url, { timeout });
        return response.data; // Equivalent to response.json() in Python
    } catch (error) {
        if (axios.isAxiosError(error)) {
            if (error.code === 'ECONNABORTED') {
                console.error(`Request timed out after ${timeout} milliseconds.`);
            } else if (error.response) {
                // HTTP error response from the server
                console.error(`HTTP error occurred: ${error.response.status} ${error.response.statusText}`);
            } else if (error.request) {
                // No response was received
                console.error('A connection error occurred.');
            } else {
                console.error(`An unexpected error occurred: ${error.message}`);
            }
        } else {
            // Non-Axios error
            console.error(`An unexpected error occurred: ${(error as Error).message}`);
        }
        throw error; // Re-throw the error for the caller to handle
    }
}
